Extract PolymorphicAction<T> from Transaction<T> & IAction

dahlia opened the pull request:
This patch extracts the subtype dispatcher which had been coupled to `Transaction<T>` and `IAction` into a new class `PolymorphicAction<T>`.  This is both a concrete `IAction` implementation and a decorator of an abstract action class having subclasses.  This also is an alternative approach to make *Libplanet.Explorer* standalone (see also #152) besides @earlbread's PR #161.

Please read added changelogs and XML comments on `IAction` and `PolymorphicAction<T>`for details.

Additionally, I renamed `InvalidActionTypeExtension` (introduced by #144) to `MissingActionTypeException`.

longfin commented on Libplanet/Tx/Transaction.cs:
```diff
@@ -584,11 +580,12 @@ internal RawTransaction ToRawTransaction(bool includeSign)
-                    { "values", a.PlainValue },
-                })
+                actions: Actions.Select(a =>
+                    a.PlainValue.ToDictionary(
```
Maybe it's better to change the type of `RawTransaction.Actions`... 😕

dahlia replied:
Currently `RawTransaction` consists of many mutable types like `byte[]` and `Dictionary<K, V>`.  If we want `RawTransaction` to be freer from state I believe we need to adjust other fields too.  Seems like it should be a separated patch.

longfin commented on Libplanet.Tests/Common/Action/DumbAction.cs:
```diff
@@ -0,0 +1,76 @@
+            RecordRehearsal = recordRehearsal;
+        }
+
+        public static ThreadLocal<ImmutableList<(Address, string)>>
```
(trivial) In the test, asynchronous processing is often performed on `Task` base without threading directly, so it is good to use [`AsyncLocal<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=netframework-4.7.2).

dahlia replied:
Applied: 7a6ebc034e102fd2f91ec9e83d4e6f0041a744a7.

longfin commented on CHANGES.md:
```diff
@@ -73,6 +73,52 @@ To be released.
+    For example, the following code:
+
+    ~~~~ csharp
+    public abstract class BaseAction : IAction { ... }
```
Can we omit `Base` prefix in docs?

dahlia replied:
It was an intention to emphasize the presence of an `abstract class` as an elaborate example.  Do you prefer simply `Action` over `BaseAction` here?

longfin replied:
I thought it would be appropriate to show as we were following the [Framework guidelines](https://stackoverflow.com/questions/429470/naming-conventions-for-abstract-classes/429494#429494). but it would be fine if it was intended.

If you intend to make it more clear, it may seem okay to attach an `Abstract` prefix.

dahlia replied:
Applied: 7a6ebc034e102fd2f91ec9e83d4e6f0041a744a7.

earlbread commented on Libplanet/Tx/Transaction.cs:
```diff
@@ -601,10 +598,8 @@ internal RawTransaction ToRawTransaction(bool includeSign)
-            var values = (IDictionary<string, object>)arg["values"];
-            action.LoadPlainValue(values.ToImmutableDictionary());
+            var action = new T();
+            action.LoadPlainValue(arg.ToImmutableDictionary());
```
Can we deserialize the actions properly without concrete action type?

dahlia replied:
Did you mean, what if `T` is not a concrete type?

earlbread replied:
Right, what if we use like `DumbAction`?

dahlia replied:
That's fine because `DumbAction` is a concrete class.  According to the [official docs][1], `new()` constraint is satisfied by only concrete class, neither `abstract class` nor `interface`:

> To use the new constraint, the type cannot be abstract.

That means, from now on we only accept a concrete class as an action type — even `PolymorphicAction<T>` in itself is concrete.  This is a point of getting rid of built-in subtype dispatcher.  Note that `T` of `PolymorphicAction<T>` does not have `new()` constraint on the other hand.  It can take an `abstract class` or an `interface`.

[1]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/new-constraint

dahlia commented on CHANGES.md:
```diff
@@ -73,6 +73,52 @@ To be released.
+    [[#169]]
+
+     -  The type parameter `T` of `Transaction<T>`, `Block<T>`, and
+        `BlockChain<T>` became to require having a parameterless constructor
```
Applied: 3c2315068acc3edf6f60958853873830a72cdf75.

dahlia commented:
I adjusted the code according @longfin's review comments, wrote more tests, and added `implicit operator PolymorphicAction<T>(T)` as well for convenience.

@longfin @earlbread Please review this again.
